model/riscv_sys_regs.sail

```diff
@@ -220,81 +219,80 @@ function get_mstatus_SXL(m : Mstatus) -> arch_xlen = {
+  if   xlen == 32 then m[SD_32]
+  else m[SD_64]
 }
 
-function legalize_mstatus(o : Mstatus, v : xlenbits) -> Mstatus = {
+function legalize_mstatus(o : Mstatus, v : bits(64)) -> Mstatus = {
   /*
    * Populate all defined fields using the bits of v, stripping anything
    * that does not have a matching bitfield entry. All bits above 32 are handled
    * explicitly later.
    */
-  // TODO: This method is error prone. We should instead explicitly copy the
-  // fields that are implemented. See legalize_mie() for an example.
-  let m : Mstatus = Mk_Mstatus(zero_extend(v[22 .. 7] @ 0b0 @ v[5 .. 5] @ 0b0 @ v[3 .. 3] @ 0b0 @ v[1 .. 1] @ 0b0));
+  let m : Mstatus = Mk_Mstatus(v);
 
+  // Set dirty bit to OR of other status bits.
+  let dirty = extStatus_of_bits(m[FS]) == Dirty |
+              extStatus_of_bits(m[XS]) == Dirty |
+              extStatus_of_bits(m[VS]) == Dirty;
+
+  [o with
+    SD_64 = if xlen == 64 then bool_to_bits(dirty) else o[SD_64],
+    SD_32 = if xlen == 32 then bool_to_bits(dirty) else o[SD_32],
+    MDT = m[MDT],
+    MPELP = m[MPELP],
+    MPV = m[MPV],
+    GVA = m[GVA],
+    /* We don't currently support changing MBE and SBE. */
+    // MBE = m[MBE],
+    // SBE = m[SBE],
+    /* We don't support dynamic changes to SXL and UXL. */
+    // SXL = if xlen == 64 then m[SXL] else o[SXL],
+    // UXL = if xlen == 64 then m[UXL] else o[UXL],
+    SDT = m[SDT],
+    SPELP = m[SPELP],
+    TSR = m[TSR],
+    TW = m[TW],
+    TVM = m[TVM],
+    MXR = m[MXR],
+    SUM = m[SUM],
+    MPRV = if extensionEnabled(Ext_U) then m[MPRV] else 0b0,
+    /* We don't have any extension context yet. */
+    XS = extStatus_to_bits(Off),
   /* FS is WARL, and making FS writable can support the M-mode emulation of an FPU
    * to support code running in S/U-modes.  Spike does this, and for now, we match it,
    * but only if Zfinx isn't enabled.
    * FIXME: This should be made a platform parameter.
    */
-  let m = [m with
-    /* Legalize MPP */
-    MPP = if have_privLevel(m[MPP]) then m[MPP] else privLevel_to_bits(lowest_supported_privLevel()),
-    /* We don't have any extension context yet. */
-    XS = extStatus_to_bits(Off),
     FS = if sys_enable_zfinx() then extStatus_to_bits(Off) else m[FS],
-    MPRV = if extensionEnabled(Ext_U) then m[MPRV] else 0b0,
-  ];
-
-  // Set dirty bit to OR of other status bits.
-  let m = [m with
-    SD = bool_to_bits(extStatus_of_bits(m[FS]) == Dirty |
-                      extStatus_of_bits(m[XS]) == Dirty |
-                      extStatus_of_bits(m[VS]) == Dirty),
-  ];
-
-  /* We don't support dynamic changes to SXL and UXL. */
-  let m = set_mstatus_SXL(m, get_mstatus_SXL(o));
-  let m = set_mstatus_UXL(m, get_mstatus_UXL(o));
-
-  /* We don't currently support changing MBE and SBE. */
-  if xlen == 64 then {
-    Mk_Mstatus([m.bits with 37 .. 36 = 0b00])
-  } else m
+    MPP = if have_privLevel(m[MPP]) then m[MPP] else privLevel_to_bits(lowest_supported_privLevel()),
+    VS = m[VS],
```

Comment:
Hmm I think we should making `FS` and `VS` read-only zero in some cases. That's an existing bug though - I made an issue: #661